### PR TITLE
Concise Object Constrcution in Csharp

### LINKED ISSUE #34: Make C# object construction more concise with object initializers
Currently we generate objects using an intermediate variable,

```csharp
GraphServiceClient graphClient = new GraphServiceClient( authProvider );

var emailAddress = new EmailAddress
{
	Address = "samanthab@contoso.onmicrosoft.com",
	Name = "Samantha Booth",
};

var attendees = new Attendee
{
	EmailAddress = emailAddress,
	Type = AttendeeType.Required,
};

var attendeesList = new List<Attendee>();
attendeesList.Add( attendees );

var location = new Location
{
	DisplayName = "Harry's Bar",
};

var end = new DateTimeTimeZone
{
	DateTime = "2017-04-15T14:00:00",
	TimeZone = "Pacific Standard Time",
};

var start = new DateTimeTimeZone
{
	DateTime = "2017-04-15T12:00:00",
	TimeZone = "Pacific Standard Time",
};

var body = new ItemBody
{
	ContentType = BodyType.Html,
	Content = "Does late morning work for you?",
};

var event = new Event
{
	Subject = "Let's go for lunch",
	Body = body,
	Start = start,
	End = end,
	Location = location,
	Attendees = attendeesList,
};

await graphClient.Me.Events
	.Request()
	.AddAsync(event);
```

however there is a more concise syntax that looks like this,

```csharp

            var event1 = new Event
            {
                Subject = "Let's go for lunch",
                Body = new ItemBody
                {
                    ContentType = BodyType.Html,
                    Content = "Does late morning work for you?",
                },
                Start = new DateTimeTimeZone
                {
                    DateTime = "2019-04-23T12:00:00",
                    TimeZone = "Pacific Standard Time",
                },
                End = new DateTimeTimeZone
                {
                    DateTime = "2019-04-23T14:00:00",
                    TimeZone = "Pacific Standard Time",
                },
                Location = new Location
                {
                    DisplayName = "Harry's Bar",
                },
                Attendees = new List<Attendee>() {
                    new Attendee
                    {
                        EmailAddress = new EmailAddress
                        {
                            Address = "samanthab@contoso.onmicrosoft.com",
                            Name = "Samantha Booth",
                        },
                        Type = AttendeeType.Required,
                    },
                }
            };
````

We should evaluate how much work it might take to generate this.

### LINKED ISSUE #43: Errors in Send Mail snippet
The send mail snippet has a couple of errors.
https://docs.microsoft.com/en-us/graph/api/user-sendmail?view=graph-rest-1.0&tabs=cs

The emailAddress is declared twice.
The variable saveToSentItems should be assigned a boolean variable and not a string.

### LINKED ISSUE #45: Escape existing double quotes present in header
Ensure that double quotes are escaped on generation of code that involves headers. 
See example [here](https://docs.microsoft.com/en-us/graph/api/user-list-events?view=graph-rest-1.0&tabs=cs#sdk-sample-code)

### PR DESCRIPTION
**Pull request by andrueastman**
This Pull requests proposes the following changes :- 

- Concise Object Constructions in C Sharp (closes #34) See example [here](https://review.docs.microsoft.com/en-us/graph/api/user-post-events?view=graph-rest-1.0&branch=andrueastman%2Fsnippets-preview&tabs=cs#sdk-sample-code)
- Escape double quotes in language specific fashion (closes #45) See example [here](https://review.docs.microsoft.com/en-us/graph/api/user-post-events?view=graph-rest-1.0&branch=andrueastman%2Fsnippets-preview&tabs=cs#sdk-sample-code)
- Repeated variables resolved as we no longer use temporary variables (closes #43) View [here](https://review.docs.microsoft.com/en-us/graph/api/user-sendmail?view=graph-rest-1.0&branch=andrueastman%2Fsnippets-preview&tabs=cs)
- Escape reserved keyword in a language specific fashion 
- Added unit tests and updated Readmes :-)

### CODE REVIEW

**Comment by finsharp** [comment]
Can the trailing command in c# object initialization be removed?
from  
```csharp
Body = new ItemBody
{
	ContentType = BodyType.Html,
	Content = "Does late morning work for you?",
},
Start = new DateTimeTimeZone
{
	DateTime = "2017-04-15T12:00:00",
	TimeZone = "Pacific Standard Time",
}
```
To this
```csharp
Body = new ItemBody
{
	ContentType = BodyType.Html,
	Content = "Does late morning work for you?"
},
Start = new DateTimeTimeZone
{
	DateTime = "2017-04-15T12:00:00",
	TimeZone = "Pacific Standard Time"
}
```

**Comment by andrueastman** [comment]
Thanks for pointing this out @finsharp. I have made updates to remove the redundant trailing commas.

File: CodeSnippetsReflection/LanguageGenerators/CSharpGenerator.cs [hunk 1]
```diff
@@ -324,13 +337,12 @@ private static string CSharpGenerateObjectFromJson(ODataPathSegment pathSegment,
...
                     //item is a primitive print as is
-                    stringBuilder.Append($"var {variableName} = {primitive};\r\n");
+                    stringBuilder.Append($"{tabSpace}{primitive}\r\n");
                     break;
             }
 
-            //add a blank line
-            stringBuilder.Append("\r\n");
-            return stringBuilder.ToString();
+            //check if this is the outermost object in a potential nested object structure and needs the semicolon termination character.
```
Comment (MIchaelMainer) [inline CodeSnippetsReflection/LanguageGenerators/CSharpGenerator.cs]: nitpick: the closing braces on either side of this comment need one more indent.

File: CodeSnippetsReflection/LanguageGenerators/CSharpGenerator.cs [hunk 2]
```diff
@@ -238,16 +248,9 @@ private static string CSharpGenerateObjectFromJson(ODataPathSegment pathSegment,
...
-                                    //new nested object needs to be constructed so call this function recursively to make it and append it before the current snippet
-                                    var newObject = CSharpGenerateObjectFromJson(pathSegment, value, newPath ,languageExpressions);
-                                    //append this at the start since it needs to be declared before this object is constructed
-                                    stringBuilder.Insert(0, newObject);
-                                    //append the usage of declared variable depending on whether its an array or simple object
-                                    stringBuilder.Append(jToken.Type == JTokenType.Object
-                                        ? $"\t{UppercaseFirstLetter(key)} = {key},\r\n"
-                                        : $"\t{UppercaseFirstLetter(key)} = {key}List,\r\n");
+                                    //new nested object needs to be constructed so call this function recursively to make it
+                                    var newObject = CSharpGenerateObjectFromJson(pathSegment, value, newPath );
```
Comment (MIchaelMainer) [inline CodeSnippetsReflection/LanguageGenerators/CSharpGenerator.cs]: Please add a test to verify that recursively generate the expected object.

File: CodeSnippetsReflection/LanguageGenerators/CSharpGenerator.cs [hunk 3]
```diff
@@ -265,52 +268,62 @@ private static string CSharpGenerateObjectFromJson(ODataPathSegment pathSegment,
...
-                                //append nested object to the list
                                 var jsonString = JsonConvert.SerializeObject(item);
-                                stringBuilder = EnsureVariableNameIsUnique(stringBuilder, path.Last());
                                 //we need to create a new object
-                                var objectItem = CSharpGenerateObjectFromJson(pathSegment, jsonString, path ,languageExpressions);
-                                //prepend the new object created before this declaration
-                                stringBuilder.Insert(0, objectItem);
-                                //add declared object to the list
-                                stringBuilder.Append($"{variableName}List.Add( {path.Last()} );\r\n");
+                                var objectStringFromJson = CSharpGenerateObjectFromJson(pathSegment, jsonString, path ).TrimEnd(";\r\n".ToCharArray());
```
Comment (MIchaelMainer) [inline CodeSnippetsReflection/LanguageGenerators/CSharpGenerator.cs]: Please add a test to assert that we correctly construct from an array.

Comment (andrueastman) [reply CodeSnippetsReflection/LanguageGenerators/CSharpGenerator.cs]: Please checkout added tests [here](https://github.com/microsoftgraph/microsoft-graph-explorer-api/pull/50/commits/da7a573dfce6fac1d6675aab29f98a5988c7b05c)

File: CodeSnippetsReflection/LanguageGenerators/CSharpGenerator.cs (hunk 2 above)
Comment (andrueastman) [reply CodeSnippetsReflection/LanguageGenerators/CSharpGenerator.cs]: Please checkout added tests [here](https://github.com/microsoftgraph/microsoft-graph-explorer-api/pull/50/commits/da7a573dfce6fac1d6675aab29f98a5988c7b05c)

**Pull request merged**